bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/openhab/core/io/transport/mqtt/internal/Subscription.java

```diff
@@ -44,8 +44,11 @@ public class Subscription {
     public void add(MqttMessageSubscriber subscriber) {
         if (subscribers.add(subscriber)) {
             // new subscriber. deliver all known retained messages
-            retainedMessages.entrySet().parallelStream()
-                    .forEach(entry -> processMessage(subscriber, entry.getKey(), entry.getValue()));
+            retainedMessages.entrySet().parallelStream().forEach(entry -> {
+                if (entry.getValue().length > 0) {
```

Comment:
Why is this necessary? Shouldn't the message always have a length? Otherwise it makes no sense to retain it.

Reply:
because we need to know that it's been retained in the past to continue retaining it in the future, even though it won't have the retained flag on future messages. so we still keep it in retainedMessages, but don't send it to new subscriptions if it's currently zero